### ISSUE TITLE
Optimisation has probably not converged to the maximum likelihood - Hessian is not positive definite

### ISSUE THREAD

**Issue by diogoroicado** [chjackson/msm#110]
Dear Dr. Jackson,

I am working on a credit card fraud detection project and encountered a challenge while estimating a multi-state model. My dataset, generated using Harris' Sparkov Data Generator ([GitHub](https://github.com/namebrandon/Sparkov_Data_Generation)), includes 50 customers with highly variable transaction counts (ranging from 3,486 to 559 per customer).

For my analysis, I defined four states: 1 - Normal behavior, 2 - Suspicious activity, 3 - Fraud and 4 - Fraud network.

Given the nature of the problem, fraudulent transactions are rare, which I suspect is affecting model estimation. Below is my dataset with all engineered features:
[data.csv](https://github.com/user-attachments/files/19299268/msm_train_data.csv)

I first attempted a simple model without covariates:
Q1.crude <- crudeinits.msm(state ~ time, subject = cc_num, data = df_train, qmatrix = Q1)
          [,1]      [,2]       [,3]     [,4]
[1,] -26.71878  26.57784   0.140940   0.0000
[2,]  38.93366 -39.07305   0.139381   0.0000
[3,]  23.46347  13.46265 -52.696648  15.7705
[4,]   8.09495   6.25519   0.000000 -14.3501
Then, I attempted to fit the model using:
msm_baseline1 <- msm(state ~ time, subject = cc_num, data = df_train, qmatrix = Q1, gen.inits = TRUE, method = "BFGS", control = list(fnscale = 500, maxit = 500, reltol = 1e-16))
However, I received the following warning:
Warning message:
In msm(state ~ time, subject = cc_num, data = df_train, qmatrix = Q1,  :
  Optimisation has probably not converged to the maximum likelihood - Hessian is not positive definite.

I have reviewed your GitHub resources and manual, adjusting parameters such as maxit, fnscale, and rescaling time, but the issue persists.

Could the severe class imbalance be causing estimation issues?
Should I modify the transition matrix initialization or impose constraints?

I would greatly appreciate any guidance or recommendations on how to improve model estimation. Thank you for your time and for developing such an invaluable tool, which has been instrumental in my research.

Best regards

**Comment by chjackson** [comment]
Make sure `obstype` or `exacttimes` is specified appropriately in the call to `msm()`.  For example if you can assume you know all the transition times, then set `exacttimes=TRUE`. 

Think about what the states actually represent (underlying status, or previous observation?) and make sure the transition structure is suitable for a continuous time model.  If the states represent an underlying status and are ordered, you should only allow transitions between adjacent states in continuous time.   For example, assuming that someone observed as "normal" on one occasion and "fraud" on the next occasion must have transitioned through "suspicious" in the middle. 

Then if the data are weak in some place (e.g. rare events) you might diagnose where the weak information is by fixing or constraining particular parameters.

**Comment by diogoroicado** [comment]
I appreciate your guidance.
I’m working with a generated dataset where multiple pre-defined profiles dictate transaction behavior, and the Python Faker library simulates all the transactions. Given this setup, I think that we can assume that all transaction timestamps are known.

From a theoretical perspective, I don’t believe transitions must strictly occur between adjacent states.  Regarding the theoretical structure of the problem, a user can make normal purchases and later be directly classified as fraudulent without necessarily passing through a suspicious stage.
However, I have established certain constraints:

- No direct transition from normal (1) or suspicious (2) to fraud network (4)
- Fraud network (4) is an absorbing state—once there, transitions back to fraud (3) do not occur
 
Does this impact the way I’m modeling the problem? Or am I missing a critical step?

By setting `exacttimes=TRUE`, the model has converged. 

`msm_baseline1 <- msm(state ~ time, subject = cc_num, data = df_train, qmatrix = Q1, gen.inits = TRUE, method = "BFGS", exacttimes=TRUE, control = list(fnscale = 500, maxit = 500, reltol = 1e-16))`

How can i optimize my models parameters?

Maximum likelihood estimates

Transition intensities
                  Baseline                    
State 1 - State 1 -26.7188 (-27.1474,-26.2969)
State 1 - State 2  26.5778 ( 26.1571, 27.0054)
State 1 - State 3   0.1409 (  0.1132,  0.1755)
State 2 - State 1  38.9337 ( 38.3173, 39.5600)
State 2 - State 2 -39.0730 (-39.7005,-38.4555)
State 2 - State 3   0.1394 (  0.1068,  0.1820)
State 3 - State 1  23.4635 ( 18.2561, 30.1563)
State 3 - State 2  13.4626 (  9.6661, 18.7504)
State 3 - State 3 -52.6966 (-62.3026,-44.5718)
State 3 - State 4  15.7705 ( 11.6121, 21.4181)
State 4 - State 1   8.0949 (  5.3301, 12.2939)
State 4 - State 2   6.2552 (  3.8886, 10.0621)
State 4 - State 4 -14.3501 (-19.6407,-10.4847)

-2 * log-likelihood:  -148899 

Now, I’d like to optimize it by:

**Adjusting Control Parameters** – Fine-tuning fnscale, maxit, and reltol
**Adding Covariates**
**Evaluating Model Performance** – Comparing different models using AIC/BIC or likelihood ratio tests
**Validating Predictions** – Checking state predictions against real transitions, possibly using pmatrix.msm() to analyze probability distributions over time

Many thanks again and I will continue to try your solutions.

**Comment by diogoroicado** [comment]
Dear Prof. Jackson,

I'm currently working on a baseline MSM model without covariates to predict state transitions based on transaction behavior. However, I have some concerns regarding the results, which seem surprisingly perfect, and I suspect there might be an underlying issue in the model setup or evaluation.

After applying the viterbi.msm() function on the validation set, I obtained a confusion matrix with 100% accuracy — all predicted states exactly match the manually classified ones:

```
predicted_states <- viterbi.msm(msm_baseline1)
viterbi_result_baseline <- viterbi.msm(msm_baseline1, newdata = df_val)
df_val$msm_baseline <- as.factor(viterbi_result_baseline$fitted)
confusionMatrix(df_val$msm_baseline, df_val$state)

          Reference
Prediction     1     2     3     4
         1 10038     0     0     0
         2     0  6942     0     0
         3     0     0    38     0
         4     0     0     0   111

Overall Statistics
                                
               Accuracy : 1     
                 95% CI : (1, 1)
    No Information Rate : 0.586 
    P-Value [Acc > NIR] : <2e-16
```

Although this outcome seems ideal, it appears too good to be true, which raises concerns about the reliability or overfitting of the model. I would greatly appreciate your guidance on the following points:

- Identical Transition Probabilities: The transition probability matrix estimated at any time t shows the same row for every state, which seems unusual:

```
pmatrix.msm(msm_baseline1, t = i)
         State 1  State 2    State 3    State 4
State 1 0.585677 0.400776 0.00468319 0.00886375
State 2 0.585677 0.400776 0.00468319 0.00886375
State 3 0.585677 0.400776 0.00468319 0.00886375
State 4 0.585677 0.400776 0.00468319 0.00886375
```

- State Sojourn Times and Transition Matrix: The average time spent in each state seems very short, especially for State 3. Additionally, the qmatrix.msm() shows some zero transitions (e.g., no transition from State 1 or 2 to State 4), which were set intentionally. Still, I wonder whether the small transition probabilities may hinder the model's generalization.
```
> qmatrix.msm(msm_baseline1)
        State 1                      State 2                      State 3                      State 4                     
State 1 -25.9507 (-26.2888,-25.6170)  25.7685 ( 25.4359, 26.1054)   0.1822 (  0.1562,  0.2127) 0                           
State 2  37.6724 ( 37.1863, 38.1649) -37.8412 (-38.3347,-37.3539)   0.1687 (  0.1390,  0.2049) 0                           
State 3  16.8099 ( 14.0664, 20.0885)  11.2529 (  9.0508, 13.9908) -37.2319 (-41.9673,-33.0309)   9.1691 (  7.2036, 11.6708)
State 4   2.4620 (  1.7314,  3.5007)   2.3825 (  1.6658,  3.4076) 0                             -4.8445 ( -6.2264, -3.7693)

> sojourn.msm(msm_baseline1)
        estimates          SE         L         U
State 1 0.0385345 0.000254505 0.0380389 0.0390366
State 2 0.0264262 0.000174725 0.0260860 0.0267709
State 3 0.0268587 0.001640654 0.0238281 0.0302747
State 4 0.2064199 0.026429358 0.1606077 0.2652997
```

- Very Low Negative Log-Likelihood: The value of -2 * log-likelihood is -221517, and the AIC is also strongly negative. Could this suggest instability in the model, or is it expected under certain conditions?

- Perfect Prediction: As mentioned above, the model perfectly reclassifies the same states defined in the training process. Could this suggest an overfitting issue or that the model is simply reproducing the logic used to define the initial states?

Could this result be due to the way the model is learning transitions from manually defined states that follow a clear, deterministic rule? Or is there something I'm missing in the model’s configuration or assumptions?

I would be truly grateful for your thoughts or any suggestions to help me troubleshoot this. I'm keen to understand why the model is overfitting to this degree, despite careful separation of training and validation data.

Thank you very much in advance for your time and support.

**Comment by chjackson** [comment]
@diogoroicado It's not clear what exactly you are trying to predict from the fitted model - I don't recognise your modelling workflow.  The Viterbi algorithm is for hidden Markov models, to impute the hidden states.  Are you using it for a non-hidden model, where the true state always equals the observed state? 

The value of the likelihood (or AIC) has no interpretation on its own.  It is for comparing between two different models.

**Comment by diogoroicado** [comment]
Initially, I had a dataset labeled with a binary outcome `is_fraud`, but I decided to expand it into four states: 1 – Normal behavior, 2 – Suspicious activity, 3 – Confirmed fraud, 4 – Fraud network involvement.

My main goal with using the msm model (e.g., `msm_baseline`, or a version including covariates) was to predict the state of a transaction based on the customer’s behavior over time. That is, I wanted to train the model on a training set and then use it to predict states on a validation set — where the states are known in both sets and then i would evaluate the models performance.

My questions are:
1. Is it possible for the msm model to give an estimate of the current state a customer is in, based on their behavior up to that point?
2. If so, I was thinking of using the output state (from the msm model)  predictions and compare it with a supervised machine learning model, to enhance predictions using the original dataset + the inferred behavioral state.
3. I even could study the use of the state from the msm model and then use a supervised model for a prediction.

Ultimately, my aim is to use the MSM model to better understand the typical progression of customer behavior before reaching a fraud state, and then improve prediction by combining this insight with supervised learning.

Thanks again for the clarification!

**Comment by chjackson** [comment]
So the person's state is unknown at the current time, but you have fitted a multistate model to a dataset that includes this person's past data?  Then you could compute $P_{rs}(t)$, the transition probabilities from their previous observed state $r$, over the time difference $t$ between the current and previous times.  You might predict the current state to be the state $s$ with the maximum transition probability.

**Comment by diogoroicado** [comment]
Yes! To train the model, I use a training set where both the true states and transaction details are known. The model learns how customer behavior changes over time.
`msm_baseline1 <- msm(state ~ time, subject = cc_num, data = df_train, qmatrix = Q1, gen.inits = TRUE, method = "BFGS", exacttimes=TRUE, control = list(fnscale = 5000, maxit = 5000, reltol = 1e-16))`

In the validation phase, I test the model by predicting the customer’s current state using only the transaction details — without using the true state.
(compute `Prs(t)`, as you suggested)

Then, I compare the predicted states to the actual states in the validation set to see how well the model performs.
If the predictions aren’t accurate enough, I plan to use the predicted state from the MSM model as an extra feature in a supervised model to help predict whether a transaction is fraudulent.

**Comment by diogoroicado** [comment]
Hello,

Following your guidance, I'm now trying to predict the next state using the following function:
```
msm_baseline1 <- msm(state ~ time, subject = cc_num, data = df_train, qmatrix = Q1, gen.inits = FALSE, method = "BFGS", exacttimes=TRUE, control = list(fnscale = 50, maxit = 500, reltol = 1e-12))

predict_state <- function(prev_state, time_diff, msm_model, covariates = NULL) {
  pmat <- pmatrix.msm(msm_model, t = time_diff, covariates = covariates)
  probs <- pmat[prev_state, ]
  return(which.max(probs))  # returns the most likely next state
}

df_val$pred_baseline <- mapply(predict_state, prev_state = df_val$prev_state, time_diff = df_val$time_diff, MoreArgs = list(msm_model = msm_baseline1))
```
However, I’m facing a problem: the transition probability matrix pmatrix.msm(msm_baseline1, t = X) returns the same values for any value of X, which causes all predictions to be identical regardless of the time difference.

Additionally, I included covariates in a new model:
`msm_top5covariate <- msm(state ~ time, subject = cc_num, data = df_train, covariates = ~ amt + time_since_last_transaction + transaction_amt_today + age + is_male, qmatrix = Q1, method = "BFGS", exacttimes=TRUE, gen.inits = TRUE, control = list(fnscale = 5000, maxit = 5000, reltol = 1e-16))`

Although the log-likelihood improved with covariates, the model’s predictions didn’t change. Can I conclude that the covariates are not influencing the predicted state transitions? (All covariates were selected using recursive feature elimination and tested for multicollinearity.)

Lastly, my training data (df_train) includes transactions within the interval [0, 𝑡]. Is it valid to use this model to predict transitions for validation data (df_val) with transactions occurring in the interval ]t, t + q]?

**Comment by chjackson** [comment]
I can't tell from where I am why your predictions are all the same.  Try to think about how they are defined mathematically as functions of other quantities, and see if you can work out why this is. 

And I can't give you any guarantee of "validity" for any particular way you are using msm.  You will have to judge whether the assumptions that the models make are reasonable in your context.

**Comment by chjackson** [comment]
Closing due to inactivity and no apparent problem in msm
